pymc3/data.py

```diff
@@ -497,10 +499,130 @@ def __new__(self, name, value):
         # transforms it to something digestible for pymc3
         shared_object = theano.shared(pm.model.pandas_to_array(value), name)
 
+        if isinstance(dims, str):
+            dims = (dims,)
+        if dims is not None and len(dims) != shared_object.ndim:
+            raise ValueError('Length of `dims` must match the dimensionality '
+                             'of the dataset.')
+
+        coords = {}
+        if isinstance(value, (pd.Series, pd.DataFrame)):
+            name = None
+            if dims is not None:
+                name = dims[0]
+            if (name is None 
+                    and value.index.name is not None
+                    and value.index.name.isidentifier()):
+                name = value.index.name
+            if name is not None:
+                coords[name] = value.index
+        if isinstance(value, pd.DataFrame):
+            name = None
+            if dims is not None:
+                name = dims[1]
+            if (name is None
+                    and value.columns.name is not None
+                    and value.columns.name.isidentifier()):
+                name = value.columns.name
+            if name is not None:
+                coords[name] = value.columns
+        if isinstance(value, np.ndarray) and dims is not None:
+            if len(dims) != value.ndim:
+                raise ValueError('Invalid data shape %s. The rank of the dataset '
+                                 'must match the length of `dims`.'
+                                 % value.shape)
+            for size, dim in zip(value.shape, dims):
+                coord = model.coords.get(dim, None)
+                if coord is None:
+                    coords[dim] = pd.RangeIndex(size, name=dim)
+
+        if export_dims:
+            model.add_coords(coords)
+
         # To draw the node for this variable in the graphviz Digraph we need
         # its shape.
         shared_object.dshape = tuple(shared_object.shape.eval())
+        if dims is not None:
+            shape_dims = model.shape_from_dims(dims)
+            if shared_object.dshape != shape_dims:
+                raise ValueError('Invalid shape. It is %s but the dimensions '
+                                 'suggest %s.'
+                                 % (shared_object.dshape, shape_dims))
+
+        model.add_random_variable(shared_object, dims=dims)
+        return shared_object
 
-        model.add_random_variable(shared_object)
 
-        return shared_object
+class _IndexAccessor:
+    def __init__(self, data):
+        self._data = data
+
+    def __getitem__(self, key):
+        category = self._data._col_as_category(key)
+        vals = self._data.data.reset_index().loc[:, key]
+        return pd.Categorical(vals, dtype=category).codes
+
+
+class TidyData:
```

Comment:
This new class is user-facing, right? In which case it probably needs some docstrings to explain what it does and how to use it. It could also be useful to update the [`pm.Data` NB](https://docs.pymc.io/notebooks/data_container.html) to show-case it? I'm happy to help with that.

Reply:
The class is prefixed with `_` and doesn't show up in `__all__`, so no, it's not user facing. A docstring would be nice though..

Reply:
I think `TidyData` is supposed to be user-facing, at least [in this example](https://user-images.githubusercontent.com/1882397/61526816-77aa5000-aa1b-11e9-82cc-840a70b5664f.png) by @aseyboldt

Reply:
This is supposed to be user-facing, but I'm not sure the implementation is good as it is. We could merge the PR without it at first and go from there.
The idea is that you can use a tidy dataframe as input, and the TidyData object automatically creates the appropriate coordinates, and mappings from observation id to the actual item.
Maybe an example of what this would do if it works correctly helps:

Say, we have a dataset like this:

```
            patient_id hospital treatment  outcome
observation                                       
0                  id1        A         C        1
1                  id2        B         C        2
2                  id3        A         D        3
3                  id1        B         D        4
```

I would like to extract from that 4 different dimensions and corresponding coordinates:
```
observation: [0, 1, 2, 3]
patient_id: ['id1', 'id2', 'id3']
hospital: ['A', 'B']
treatment: ['C', 'D']
```
And also mappings, that allow us to index a array with dim=treatment etc to get one with dim=observation. So that we can run linear regressions like this:
```
mu = (
    intercept
    + patient_effect[obs_to_patient_idx]
    + hospital_effect[obs_to_hospital_idx]
    + treatment_effect[obs_to_treatment_idx]
)
```
where
```
obs_to_patient_idx = [0, 1, 2, 0]
obs_to_hostpital_idx = [0, 1, 0, 1]
obs_to_treatment_idx = [0, 0, 1, 1]
```
Where mu has now dim=observation.